mllib/src/main/scala/org/apache/spark/ml/regression/RandomForestRegressor.scala

```diff
@@ -168,15 +173,39 @@ class RandomForestRegressionModel private[ml] (
   // Note: We may add support for weights (based on tree performance) later on.
   private lazy val _treeWeights: Array[Double] = Array.fill[Double](_trees.length)(1.0)
 
+  @Since("2.1.0")
+  /** @group getParam */
+  def setVarianceCol(value: String): this.type = set(varianceCol, value)
+
   @Since("1.4.0")
   override def treeWeights: Array[Double] = _treeWeights
 
   override protected def transformImpl(dataset: Dataset[_]): DataFrame = {
     val bcastModel = dataset.sparkSession.sparkContext.broadcast(this)
+
+    var output = dataset
+
     val predictUDF = udf { (features: Any) =>
       bcastModel.value.predict(features.asInstanceOf[Vector])
     }
-    dataset.withColumn($(predictionCol), predictUDF(col($(featuresCol))))
+    val predictions = predictUDF(col($(featuresCol)))
+    output = dataset.withColumn($(predictionCol), predictions)
+
+    val varianceUDF = udf { (features: Any) =>
+      val leafNodes = bcastModel.value.returnLeafNodes(features.asInstanceOf[Vector])
+      leafNodes.map { leafNode =>
```

Comment:
Nice! I'll address this for the `RandomForest` here and we can move the decision tree test strengthening to another PR.